### ISSUE TITLE
# [Perf] XLA fails to fuse segmented reductions in softmax (segment_max → exp → segment_sum), 8-19x slower than hand-fused Triton

### ISSUE THREAD

**Issue by Umang-projects** [jax-ml/jax#37511]
### Problem
JAX's segmented softmax (via `jax.ops.segment_max` + `jax.ops.segment_sum`) compiles to **3 separate GPU kernels** because XLA cannot fuse `stablehlo.scatter` operations with intervening elementwise ops. A hand-written Triton kernel achieves **8-19x speedup** by fusing everything into a single SRAM pass.

### Benchmark (A100 / A10 / T4)
| Segments | Length | JAX (ms) | PyTorch Geometric (ms) | Triton Fused (ms) | Triton vs JAX |
|----------|--------|----------|------------------------|-------------------|---------------|
| 10,000   | 64     | 0.288    | 0.488                  | 0.033             | **8.78x**     |
| 10,000   | 128    | 0.431    | 0.629                  | 0.032             | **13.58x**    |
| 10,000   | 256    | 0.869    | 0.478                  | 0.035             | **24.97x**    |
| 50,000   | 128    | 1.455    | 0.880                  | 0.132             | **11.01x**    |

### Root Cause (HLO Evidence)
Dumped HLO shows `stablehlo.scatter` creates hard fusion boundaries:

```mlir
// 1. segment_max → separate kernel + HBM write
%2 = "stablehlo.scatter"(%0, %1, %arg0) <{...}> ({
  ^bb0(%arg2: tensor<f32>, %arg3: tensor<f32>):
    %23 = stablehlo.maximum %arg2, %arg3 : tensor<f32>
    stablehlo.return %23 : tensor<f32>
})

// 2. exp → HBM read, compute, HBM write (NOT fused with step 1)
%11 = stablehlo.exponential %10 : tensor<1280000xf32>

// 3. segment_sum → separate kernel + HBM write (NOT fused with step 2)
%14 = "stablehlo.scatter"(%12, %13, %11) <{...}> ({
  ^bb0(%arg2: tensor<f32>, %arg3: tensor<f32>):
    %23 = stablehlo.add %arg2, %arg3 : tensor<f32>
    stablehlo.return %23 : tensor<f32>
})

// 4. divide → HBM read + compute (NOT fused)
%22 = stablehlo.divide %11, %21 : tensor<1280000xf32>
```

### Reproduction
```python
import jax
import jax.numpy as jnp

@jax.jit(static_argnames=['num_segments'])
def jax_segmented_softmax(scores, segment_ids, num_segments):
    max_val = jax.ops.segment_max(scores, segment_ids, num_segments=num_segments)
    scores_shifted = scores - max_val[segment_ids]
    exp_scores = jnp.exp(scores_shifted)
    sum_exp = jax.ops.segment_sum(exp_scores, segment_ids, num_segments=num_segments)
    return exp_scores / sum_exp[segment_ids]
```

### Environment
- JAX: `jax[cuda12]` 
- GPU: A100 / A10G / T4
- XLA: StableHLO via `lower().as_text()`

### Questions / Proposal
1. Is there ongoing work to enable **fusion across matching scatter indices** for reductions?
2. Alternatively, could `segment_max` + `segment_sum` be lowered to a single fused op when indices match?

### Proposed Solution / Request
Segmented Softmax is a massive bottleneck for Graph Neural Networks (e.g., GATs). Since scatter fusion seems to be a known limitation in XLA, would the JAX team be open to a PR implementing a custom `jax.experimental.pallas` kernel for this?

I have the SRAM-fused logic ready and verified in Triton, and I would love to port this to Pallas to bridge this massive performance gap natively within JAX.

### System info (python version, jaxlib version, accelerator, etc.)

jax:    0.4.28
jaxlib: 0.4.28
python: 3.12.3
platform: Linux-x86_64
Accelerator: NVIDIA A100

**Comment by Umang-projects** [comment]
@sharadmv @mattjj Would love to get your thoughts and review.
